src/nbl/ext/EnvmapImportanceSampling/EnvmapImportanceSampling.cpp

```diff
@@ -0,0 +1,426 @@
+			copyRegion.imageOffset = { 0u,0u,0u };
+			const auto extent = lumaImage->getMipSize(copyRegion.imageSubresource.mipLevel);
+			copyRegion.imageExtent = { extent.x,extent.y,extent.z };
+		}
+		const uint32_t lastMipTexelCount = copyRegion.imageSubresource.layerCount*copyRegion.imageExtent.depth*copyRegion.imageExtent.height*copyRegion.imageExtent.width;
+		const uint32_t size = lastMipTexelCount*asset::getTexelOrBlockBytesize(lumaImage->getCreationParameters().format);
+
+		// remember that without initializing the address to be allocated to invalid_address you won't get an allocation!
+		auto downloadStagingArea = m_driver->getDefaultDownStreamingBuffer();
+		uint32_t address = std::remove_pointer<decltype(downloadStagingArea)>::type::invalid_address;
+		// allocate
+		{
+			// common page size
+			const uint32_t alignment = 4096u;
+			const auto waitPoint = std::chrono::high_resolution_clock::now()+std::chrono::nanoseconds(timeoutInNanoSeconds);
+			auto unallocatedSize = downloadStagingArea->multi_alloc(waitPoint,1u,&address,&size,&alignment);
+			if (unallocatedSize)
+			{
+				os::Printer::log("Could not download the last luma mip map level from the GPU!", ELL_ERROR);
+				return core::nan<float>();
+			}
+		}
+
+		//
+		copyRegion.bufferOffset = address;
+		m_driver->copyImageToBuffer(lumaImage.get(),downloadStagingArea->getBuffer(),1,&copyRegion);
+
+		// place and wait for download fence
+		{
+			auto downloadFence = m_driver->placeFence(true);
+			auto result = downloadFence->waitCPU(timeoutInNanoSeconds,true);
+			//
+			if (result==E_DRIVER_FENCE_RETVAL::EDFR_TIMEOUT_EXPIRED || result==E_DRIVER_FENCE_RETVAL::EDFR_FAIL)
+			{
+				os::Printer::log("Could not download the last luma mip map level from the GPU! Fence not Signalled!", ELL_ERROR);
+				downloadStagingArea->multi_free(1u,&address,&size,nullptr);
+				return core::nan<float>();
+			}
+			// then invalidate the CPU cache of the mapping
+			if (downloadStagingArea->needsManualFlushOrInvalidate())
+				m_driver->invalidateMappedMemoryRanges({ {downloadStagingArea->getBuffer()->getBoundMemory(),address,size} });
+		}
+
+		//
+		{
+			const float* r32fData = reinterpret_cast<float*>(reinterpret_cast<uint8_t*>(downloadStagingArea->getBufferPointer())+address);
+			const auto avgVal = std::reduce(r32fData,r32fData+lastMipTexelCount)/float(lastMipTexelCount);
+			pdfNormalizationFactor = 1.0/(2.0*core::PI<double>()*core::PI<double>()*avgVal);
+		}
+		downloadStagingArea->multi_free(1u,&address,&size,nullptr);
+	}
+
+	// Generate WarpMap
+	{
+		m_driver->bindComputePipeline(m_warpPipeline.get());
+		m_driver->bindDescriptorSets(EPBP_COMPUTE,m_warpPipeline->getLayout(),0u,1u,&m_warpDS.get(),nullptr);
+		m_driver->dispatch(m_warpWorkgroups[0],m_warpWorkgroups[1],1);
+		COpenGLExtensionHandler::pGlMemoryBarrier(GL_ALL_BARRIER_BITS); // TODO: rethink when reimplementing in Vulkan
+	}
+
+	return enableRIS;
+}
+
```

Comment:
@kevyuu code from the old ancient `ditt` branch..

BTW we need the envmap to be now usable with and without RIS (MIS instead).